Retry / File State Storage

vaibhav-db opened the pull request:
refactor retry handler

vaibhav-db committed:
SYMPHONYP-940 MS Teams feature to submit the adaptive card

vaibhav-db committed:
Pull request #29: SYMPHONYP-940 MS Teams feature to submit the adaptive card

Merge in SYMPHONYP/symphony-java-toolkit from feature/SYMPHONYP-939-ms-teams-spring-bot-retry-messages to spring-bot-master-db

* commit '3f2fec632e13056b0fec79bffe62697fec0e4913':
  SYMPHONYP-940 MS Teams feature to submit the adaptive card

vaibhav-db committed:
SYMPHONYP-940 refactor retry handler

vaibhav-db committed:
Merge branch 'spring-bot-master' of https://stash.gto.intranet.db.com:8082/scm/symphonyp/symphony-java-toolkit into feature/SYMPHONYP-939-ms-teams-spring-bot-retry-messages

vaibhav-db committed:
SYMPHONYP-939 implements review comments

abhishekdahiya-db committed:
added testcases for Message Retry.

abhishekdahiya-db committed:
review comments changes done.

vaibhav-db committed:
Pull request #30: SYMPHONYP-940 refactor retry handler

Merge in SYMPHONYP/symphony-java-toolkit from feature/SYMPHONYP-939-ms-teams-spring-bot-retry-messages to spring-bot-master-db

* commit '3e4b60cbfa4318e45122d113b408aec181c8bf98':
  review comments changes done.
  added testcases for Message Retry.
  SYMPHONYP-939 implements review comments
  SYMPHONYP-940 refactor retry handler
  fixing formatting
  Adding roadmap
  Update LICENSE
  setting copyright name

robmoffat commented on libs/teams/teams-chat-workflow-spring-boot-starter/src/main/java/org/finos/springbot/teams/handlers/TeamsResponseHandler.java:
```diff
@@ -62,7 +61,7 @@ public class TeamsResponseHandler implements ResponseHandler, ApplicationContext
 	protected TeamsStateStorage teamsState;
 	protected TeamsConversations teamsConversations;
-	protected MessageRetryHandler messageRetryHandler;
+	protected MessageRetryHandler retryHandler;
```
can we not have retryHandler in a subclass?  e.g. `RetryingTeamsResponseHandler`?

pankaj-a-khandelwal-db replied:
Changes done as per review comments

robmoffat commented on libs/teams/teams-chat-workflow-spring-boot-starter/src/main/java/org/finos/springbot/teams/handlers/TeamsResponseHandler.java:
```diff
@@ -142,7 +141,7 @@ protected TemplateType getTemplateType(WorkResponse wr) {
 		if (displayTemplater.hasTemplate(wr)) {
 			tt = TemplateType.THYMELEAF;
-		} else if (workTemplater.hasTemplate(wr)) {
+		} else if (workTemplater.hasTemplate(wr) || AdaptiveCardPassthrough.isAdaptiveCard(wr)) {
```
it's odd that this has changed?

vaibhav-db replied:
Hi Rob,

We have usecase where we wanted to send adaptive card as passthrough like user send Adaptive card json as string input parameter and bot will send this adaptive card to ms teams channel.

robmoffat commented on libs/teams/teams-chat-workflow-spring-boot-starter/src/main/java/org/finos/springbot/teams/handlers/TeamsResponseHandler.java:
```diff
@@ -274,7 +251,7 @@ public void retryMessage() {
 
 		Optional<MessageRetry> opt;
-		while ((opt = messageRetryHandler.get()).isPresent()) {
+		while ((opt = retryHandler.get()).isPresent()) {
```
please make sure we have code coverage of this.

abhishekdahiya-db committed:
bug fix implemented.

vaibhav-db committed:
refactor retry message handler

vaibhav-db committed:
refactor retry message handler

vaibhav-db committed:
refactor retry message handler

abhishekdahiya-db committed:
Pull request #31: bug fix implemented.

Merge in SYMPHONYP/symphony-java-toolkit from bugfix/SYMPHONYP-946-History-Api-bug-fix to spring-bot-master-db

* commit '18d766fd56c9bb561b2de3179c75cd57fae5e3eb':
  bug fix implemented.

vaibhav-db committed:
refactor retry message handler

vaibhav-db committed:
refactor retry message handler

vaibhav-db committed:
refactor retry message handler

vaibhav-db committed:
Pull request #32: Feature/SYMPHONYP-947 teams spring bot retry handler refactor code

Merge in SYMPHONYP/symphony-java-toolkit from feature/SYMPHONYP-947-teams-spring-bot-retry-handler-refactor-code to spring-bot-master-db

* commit '320488b2b314041fab0498f244763a1c2250973c':
  refactor retry message handler
  refactor retry message handler
  refactor retry message handler
  refactor retry message handler
  refactor retry message handler
  refactor retry message handler

vaibhav-db committed:
Fixed state storage testcases

robmoffat commented on libs/teams/teams-chat-workflow-spring-boot-starter/src/test/java/org/finos/springbot/teams/state/AbstractStateStorageTest.java:
```diff
@@ -72,6 +72,28 @@ public void testStoreWithTagDates() throws IOException {
 	}
 	
+	@Test
+	public void testSlashStoreWithMultTags() throws IOException {
```
testFilenamesWithSlash

robmoffat commented on libs/teams/teams-chat-workflow-spring-boot-starter/src/test/java/org/finos/springbot/teams/state/AbstractStateStorageTest.java:
```diff
@@ -72,6 +72,28 @@ public void testStoreWithTagDates() throws IOException {
+		tss.store("thefile/thefile", tags1, somedata);
+		tss.store("thefile/theotherfile", tags2, somedata);
+	
+		Assertions.assertEquals(1, hoover(tss.retrieve(tagList1, false)).size());	
```
shouldn't you also check you can retrieve the other file as well?

vaibhav-db replied:
ok.. I will add other file also.

robmoffat commented on libs/teams/teams-chat-workflow-spring-boot-starter/src/main/java/org/finos/springbot/teams/handlers/retry/InMemoryRetryingActivityHandler.java:
```diff
@@ -0,0 +1,37 @@
+
+public class InMemoryRetryingActivityHandler extends AbstractRetryingActivityHandler  {
+
+	private Queue<MessageRetry> queue = new ConcurrentLinkedQueue<>();
```
Have you looked at this?  https://stackoverflow.com/questions/40485398/retry-logic-with-completablefuture

vaibhav-db replied:
I have already tried this, it was not working. As per our requirement if exception occurred we need to wait for Retry-After then try retry.

robmoffat replied:
What didn't work about it?  Is it the lack of delay?  I am looking at this code from the SO page:

```
public CompletableFuture<Result> executeActionAsync() {
    CompletableFuture<Result> f=executeMycustomActionHere();
    for(int i=0; i<MAX_RETRIES; i++) {
        f=f.thenApply(CompletableFuture::completedFuture)
           .exceptionally(t -> executeMycustomActionHere())
           .thenCompose(Function.identity());
    }
    return f;
}

```

robmoffat replied:
should not be too hard to add a delay on the .exceptionally() line, if that is the problem

vaibhav-db replied:
No our requirement is to wait for Retry-Afer time, but we can't add the sleep or wait as it will wait the thread. we have implemented similar only inserted data into queue. and added cron job to resent the data.

robmoffat replied:
but we have this code to get the retry after time:

```
public long getRetryAfter(CompletionException e) {
		ErrorResponseException ere = (ErrorResponseException) ((CompletionException) e).getCause();
		retrofit2.Response<ResponseBody> response = ere.response();
		String retryAfter = response.headers().get("Retry-After");

		long retryAfterInt = 1;// initiate to 1 sec
		if (StringUtils.isNumeric(retryAfter)) {
			retryAfterInt = Long.parseLong(retryAfter);
		}

		return retryAfterInt;
	}
```


So couldn't you just have some code like this:

```
@Override
	public CompletableFuture<ResourceResponse> handleActivity(Activity activity, TeamsAddressable to) {
		CompletableFuture<ResourceResponse> f= tc.handleActivity(activity, to);
	    for(int i=0; i<teamsRetryCount; i++) {
	        f=f.thenApply(CompletableFuture::completedFuture)
	           .exceptionally(t -> {
	        	   if (isTooManyRequest(t)) {
	        		   long ra = getRetryAfter((CompletionException) t);
	        		   Thread.sleep(ra);
	        		   return  tc.handleActivity(activity, to);
	        	   } else {
	        		   return CompletableFuture.failedFuture(t);
	        	   }
	        	   
	           })
	           .thenCompose(Function.identity());
	    }
	    return f;
	}

```

Apologies if you already tried this - I'm also trying to get my head around `CompletableFuture`s!

vaibhav-db replied:
we just don't want to used  Thread.sleep(ra); in code. As it will keep the current thread on wait. :)

We can connect on Monday, I will explain you.
I am learning the CompletableFuture :)

robmoffat replied:
argh you're right!! 😦 

How about this: https://github.com/deutschebank/spring-bot/pull/1

I actually wrote a test here - well, changed your `InMemoryRetryingActivityHandlerTest` to see if this actually worked.

Test passes but maybe you have more to add..?

vaibhav-db replied:
Hi Rob,

Testcase only pass, if we have `for (int i = 0; i < teamsRetryCount; i++) {` for loop in code. But in our case, if exception occurred we simply insert that record in in-memory queue. 
Here we have to check the how to write testcase around schedule also.

robmoffat replied:
Hi, we can wait until Wednesday to talk about this or we can maybe have a call a bit sooner?

vaibhav-db replied:
Hi Rob, we will connect tomorrow. Completely overload with teams.
I have tried CompletableFuture.delayedExecutor(ra, TimeUnit.MILLISECONDS) but it is in java 9, so we can't go for it. 
Alternate solution i found as scheduler.
https://stackoverflow.com/questions/58707689/is-it-possible-to-schedule-a-completablefuture

vaibhav-db committed:
State storage testcases.

vaibhav-db committed:
refactor file stage storage and testcase

vaibhav-db committed:
Teams retry InMemory handler

vaibhav-db committed:
Teams retry InMemory handler

robmoffat reviewed (approved):
looks good to me

vaibhav-db committed:
Corrected InMemoryRetryingActivityHandler reference

vaibhav-db committed:
Error stream id in log
